[CodeComplete] Fix code completion of initialization for variable wrapped by generic property wrapper

ahoppen opened the pull request:
If completing the initialization of a variable that’s wrapped in a generic, unbound property wrapper, the expression's type is an `UnboundGenericType`. AFAICT that’s expected and the correct type to assign.

We hit the first assertion failure by trying to retrieve that type's substitution map. `UnboundGenericType`s were not handled here, so we hit an assertion. AFAICT we can't get any substitutions out of an unbound generic type, so we should just continue looking into the parent type.

We hit the second assertion when trying to retrieve the property wrapper’s backing type, which is null (becuase it couldn't be resolved). However, we haven’t emitted a diagnostic because the variable declaration is perfectly valid. Hence I’m removing the assertion.

Fixes rdar://64141399

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on lib/Sema/TypeCheckPropertyWrapper.cpp:
```diff
@@ -580,7 +580,6 @@ PropertyWrapperBackingPropertyTypeRequest::evaluate(
 
   ASTContext &ctx = var->getASTContext();
   Type type = ctx.getSideCachedPropertyWrapperBackingPropertyType(var);
-  assert(type || ctx.Diags.hadAnyError());
```
Instead of removing this assertion, could you insert
```
    if (binding->isInvalid())
      return Type();
```
at the end of `if (binding && binding->isInitialized(index))` branch?

ahoppen replied:
Ah, that looks like a cleaner solution. Applied it.

rintaro commented on test/SourceKit/CodeComplete/complete_generic_property_wrapper_initialization.swift:
```diff
@@ -0,0 +1,32 @@
+// RUN: %complete-test %s -tok=COMPLETE_GENERIC
```
`complete-test` is SourceKit test. For non-SourceKit specific changes we use `swift-ide-test` in `test/IDE/complete_*.swift`. But in this specific case, please use `validation-test/IDE/crashers_2_fixed/`.

ahoppen replied:
Ah, didn’t know about `validation-test/IDE/crashers_2_fixed`. Moved it there.

rintaro commented on test/SourceKit/CodeComplete/complete_generic_property_wrapper_initialization.swift:
```diff
@@ -0,0 +1,32 @@
+
+public struct MyStruct {
+  @GenericWrapper var someProperty = #^COMPLETE_GENERIC^#
+  @GenericContext<Int>.GenericWrapper var someProperty2 = #^COMPLETE_IN_GENERIC_CONTEXT^#
```
When testing `COMPLETE_IN_GENERIC_CONTEXT`, other tokens are just erased from the source so,
```swift
   @GenericWrapper var someProperty = #^COMPLETE_GENERIC^#
   @GenericContext<Int>.GenericWrapper var someProperty2 = #^COMPLETE_IN_GENERIC_CONTEXT^#
```
is (probably) parsed as
```swift
@GenericWrapper var someProperty = @GenericContext<Int>.GenericWrapper /* missing ';' */ var someProperty2 = <TOKEN>
```
To prevent this, please wrap them separate types.

ahoppen replied:
👍 Moved them to two different types. Didn’t think about that.

rintaro commented:
@slavapestov could you review `getContextSubstitutions()` change?

ahoppen committed:
[CodeComplete] Fix code completion of initialization for variable wrapped by generic property wrapper

If completing the initialization of a variable that’s wrapped in a generic, unbound property wrapper, the expression's type is an `UnboundGenericType`. AFAICT that’s expected and the correct type to assign.

We hit the first assertion failure by trying to retrieve that type's substitution map. `UnboundGenericType`s were not handled here, so we crashed. AFAICT we can't get any substitutions out of an unbound generic type, so we should just continue looking into the parent type.

We hit the second assertion when trying to retrieve the property wrapper’s backing type, which is null (becuase it couldn't be resolved). However, we haven’t emitted a diagnostic because the variable declaration is perfectly valid. Hence I’m removing the assertion.

Fixes rdar://64141399

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test macOS
